packages/eslint-plugin/src/rules/no-unnecessary-condition.ts

```diff
@@ -63,13 +63,83 @@ const isPossiblyNullish = (type: ts.Type): boolean =>
 const isAlwaysNullish = (type: ts.Type): boolean =>
   tsutils.unionTypeParts(type).every(isNullishType);
 
-// isLiteralType only covers numbers and strings, this is a more exhaustive check.
-const isLiteral = (type: ts.Type): boolean =>
-  tsutils.isBooleanLiteralType(type) ||
-  type.flags === ts.TypeFlags.Undefined ||
-  type.flags === ts.TypeFlags.Null ||
-  type.flags === ts.TypeFlags.Void ||
```

Comment:
Note that I intentionally omitted the `void` check.

1. It is not tested for
2. We've gone towards thinking of `void` as `unknown` in terms of its possible values, rather than `undefined` only. And if it doesn't have a specific, knowable value such that we can precompute the condition... then it's an indication that we can't in good faith tell the user the condition is unnecessary.